src/vsl/parser/parser.ne

```diff
@@ -0,0 +1,458 @@
+Prefix
+   -> ("-" | "+" | "*" | "**" | "!" | "~") Prefix {%
+        (data, location) =>
+            new t.UnaryExpression(data[1], data[0][0].value, location)
+    %}
+    | Property {% id %}
+
+# ============================================================================ #
+#                                  Properties                                  #
+# ============================================================================ #
+
+@{%
+function recursiveProperty(head, tail, optional, location) {
+    if (tail.length === 0) {
+        return head;
+    }
+    for (let i = 0; i < tail.length - 1; i++) {
+        tail[i + 1].head = tail[i];
+    }
+    tail[0].head = head;
+    tail[tail.length - 1].optional = optional;
+    //console.log('ok', require('util').inspect(tail[tail.length - 1],
+    //    {depth: null}));
+    return tail[tail.length - 1];
+}
+%}
+
+# Properties
+Property
+   -> propertyHead (_ propertyTail {% nth(1) %}):* {%
+        (data, location) => {
+            if (data[1].length === 0) {
+                return data[0];
+            } else {
+                data = recursiveProperty(data[0], data[1], location);
+                return data;
+            }
+        }
+    %}
+    | "?" (_ nullableProperty {% nth(1) %}):* {%
+        (data, location) => {
+            if (data[1].length === 0) {
+                return new t.Whatever(location);
+            } else {
+                return recursiveProperty(new t.Whatever(location), data[1],
+                    location);
+            }
+        }
+    %}
+
+propertyHead
+   -> Identifier             {% id %}
+    | "(" _ Expression _ ")" {% nth(2) %}
+    | FunctionizedOperator   {% id %}
+    | Literal                {% id %}
+
+## TODO: do we include short-circuit (&& and ||)? if so, how to implement?
+## what about assignment
+FunctionizedOperator
+   -> "(" ("==" | "!=" | "<>" | "<=>" | "<=" | ">=" | ">" | "<" | "<<" | ">>" |
+        "+" | "-" | "*" | "/" | "**" | "&" | "|" | "^" | "~>" | ":>" | ".." |
+        "..." | "::") ")" {%
+        (data, location) =>
+            new t.FunctionizedOperator(data[1][0].value, location)
+    %}
+
+propertyTail
+   -> "." _ Identifier {%
+        (data, location) =>
+            new t.PropertyExpression(null, data[2], false, location)
+    %}
+    | "[" _ (delimited[Expression, "," _] {% id %}) _ "]" {%
+        (data, location) => new t.Subscript(null, data[2], false, location)
+    %}
+    | nullableProperty {% id %}
+
+nullableProperty
+   -> "?" "." _ Identifier {%
+        (data, location) =>
+            new t.PropertyExpression(null, data[3], true, location)
+    %}
+    | "?" "[" _ (delimited[Expression, "," _] {% id %}) _ "]" {%
+        (data, location) => new t.Subscript(null, data[3], true, location)
+    %}
+    | "(" _ FunctionCallList _ ")" {% nth(2) %}
+
+FunctionCallList
+   -> delimited[FunctionCallArgument {% id %}, _ "," _]:? {%
+        (data, location) => new t.FunctionCall(null, data[0], location)
```

Comment:
Since you have `:?` I believe it should be `data[0] || []` as `data[0]` would be null whereas it should be an array